tests/python/pants_test/engine/test_rules.py

```diff
@@ -244,18 +274,22 @@ def d_from_a(a: A) -> D:
     with self.assertRaises(Exception) as cm:
       create_scheduler(rules)
 
-    self.assert_equal_with_printing(dedent("""
-                     Rules with errors: 3
-                       (A, [B, C], a_from_b_and_c()):
-                         Was not usable by any other @rule.
-                       (A, [C, B], a_from_c_and_b()):
-                         Was not usable by any other @rule.
-                       (D, [A], d_from_a()):
-                         Ambiguous rules to compute A with parameter types (B+C):
-                           (A, [B, C], a_from_b_and_c()) for (B+C)
-                           (A, [C, B], a_from_c_and_b()) for (B+C)
-                     """).strip(),
-      str(cm.exception))
+    self.assert_equal_with_printing(
+      dedent(
+        f"""\
+        Rules with errors: 3
+          {fmt_rule(a_from_b_and_c)}:
+            Was not reachable, either because no rules can produce the params or it was shadowed by another @rule.
```

Comment:
What does "shadowed" mean here? It's not immediately obvious, I think.

Reply:
Honestly, not very certain. I took that language from comments in Rust and the test `test_unreachable_rule`. I'd love to improve this if anyone has recommendations!